xmppserver/src/main/java/org/jivesoftware/openfire/net/SocketReader.java

```diff
@@ -461,8 +463,12 @@ private boolean isHostUnknown(String host) {
             // requested host matched the server name
             return false;
         }
-        // Check if the host matches a subdomain of this host
-        return !routingTable.hasComponentRoute(new JID(host));
+        if (routingTable.hasComponentRoute(new JID(host))) {
+            // Check if the host matches a subdomain of this host
+            return false;
+        }
+
+        return !JiveGlobals.getListProperty("xmpp.gateway.domains", Collections.emptyList()).contains(host);
```

Comment:
How would this work when the stanza is for a service domain, e.g. conference or pubsub?

Reply:
Interesting thought. This is something that we should test for.

The problem might also pop up when resolving the IP address for the remote XMPP host that is serving that component.

I assume that most setups will not have DNS entries for those component domains (although they might have, when they do trickery to get Let's Encrypt certificates generated!)

This is what I think will mostly happen: When there are no DNS SRV or A entries for the component domains, then the lookup for XMPP-service-to-Server-hostname should fall back from using the XMPP domain of `conference.example.org` (no DNS SRV record) to the 'parent' XMPP domain of `example.org` (which does have a DNS SRV record), which then resolves to a hostname, and ultimately, an IP address.

When there are DNS entries for the component domains (A and/or SRV), then I'm not sure what happens.

Should we require administrators to add those to the `xmpp.gateway.domains` and/or `dns.override`, or can we, safely, facilitate things in code?

Reply:
Openfie used to, years ago, "walk the tree" to find DNS SRV records, but that has various risks when a parent domain injects records. If there's no conference.example.org DNS records, most servers won't connect to it. That's OK.